draft-ietf-ppm-dap.md

```diff
@@ -1099,84 +1099,170 @@ refinement process. We instead think of these as properties of the output
+type: one of "initialize", "continue", "finish", or "finished". The type of
+each step, and how many steps there are, depends on the VDAF. The message
+structure and processing rules are specified in {{preparation-messages}}.
+
+In general, reports cannot be aggregated until the Collector specifies an
+aggregation parameter. However, in some situations it is possible to begin
+aggregation as soon as reports arrive. For example, Prio3 has just one valid
+aggregation parameter (the empty string). And there are use cases for Poplar1
+in which aggregation can begin immediately (i.e., those for which the candidate
+prefixes/strings are fixed in advance).
+
+The aggregation flow can be thought of as having three phases:
+
+- Initialization: Begin the aggregation flow by disseminating report shares and
+  initializing the VDAF preparation state for each report.
+- Continuation: Continue the aggregation flow by exchanging prep shares and
+  messages until preparation completes or an error occurs.
+- Completion: Finish the aggregate flow, yielding an output share corresponding
+  to each report report share in the batch.
+
+These phases are described in {{agg-init}} and {{agg-continue-flow}}.
+
+### Preparation Messages
+
+[TODO: Replace definitions of ReportShareError, PrepareStepState, and
+PrepareStep with backward references to this section.]
+
+At any point in the preparation process, a report can be rejected by one of the
+Aggregators and removed from the aggregation job. When a report is rejected, it
+is replaced by a `ReportShareError`:
+
+~~~ message
+enum {
+  batch_collected(0),
+  report_replayed(1),
+  report_dropped(2),
+  hpke_unknown_config_id(3),
+  hpke_decrypt_error(4),
+  vdaf_prep_error(5),
+  batch_saturated(6),
+  task_expired(7),
+  unrecognized_message(8),
+  report_too_early(9),
+  (255)
+} ReportShareError;
+~~~
+
+As shown in Figure 6 of {{!VDAF}}, the preparation process is organized into a
+sequence of rounds. At the start of each round, each Aggregator produces a
+value called a "preparation message-share", or "prep share" for short. The prep
+shares are combined into the preparation message, or "prep message", which is
+used as input for the next round. (The value output in the final round is the
+Aggregator's output share.)
+
+Each step of the aggregation sub-protocol is encoded as a `PrepareStep` message
+as follows:
+
+~~~ message
+enum {
+  initialize(0),
+  continue(1),
+  finish(2),
+  finished(3),
+  reject(4),
+  (255)
+} PrepareStepState;
+
+struct {
+  PrepareStepState prepare_step_state;
+  select (PrepareStep.prepare_step_state) {
+    case initialize:
+      opaque prep_share<0..2^32-1>;
+    case continue:
+      opaque prep_msg<0..2^32-1>;
+      opaque prep_share<0..2^32-1>;
+    case finish:
+      opaque prep_msg<0..2^32-1>;
+    case finished:
+      Empty;
+    case reject:
+      ReportShareError report_share_error;
+  };
+} PrepareStep;
```

Comment:
I still don't understand why this information has to be surfaced at this level.

Given that this message is self-contained, why can't this message be what's passed to the VDAF and let it figure out what to do. Please don't tell me that that's how the VDAF API works, because I am suggesting a different API.

Reply:
My understanding for the reason for this complexity is:
* DAP is moving towards specifying to exactly two aggregators (this work is part of that effort).
* VDAF is currently specified to an arbitrary number of aggregators.

... so somewhere we need to specify how to adapt from the arbitrary-aggregator VDAF interface to the 2-aggregator DAP protocol, in a way that unlocks the performance benefit allowed by 2-aggregator DAP. This PR contains that complexity, and attempts to encapsulate it into the DAPPrepInit/DAPPrepNext functions (replacing the similar VDAF-interface VDAF.prep_init & VDAF.prep_next functions).

If we want to avoid this complexity entirely in DAP, the most likely change would be to the VDAF specification, either to specify VDAF to 2 aggregators or to specify a second VDAF interface which is equivalent to the 2-aggregator DAPPrepInit/DAPPrepNext interface described in this PR. I'm less confident to speak to changes to VDAF, but VDAF editors have expressed a desire to maintain that VDAF can work with an arbitrary number of aggregators. Perhaps we should consider specifying the functionality of this PR in VDAF? (I do not hold a strong opinion about where the better place for this complexity is, I'm just highlighting that that seems to be the tradeoff.)

Reply:
It's easy enough to move the wrapper code to VDAF, I'll create a PR there today.

Reply:
https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/240